src/nodeattr/nodeattr.c

```diff
@@ -1224,6 +1240,265 @@ compress(genders_t gp)
+
+        /* multiple empty attrs will result in multiple commas,
+         * I don't think that's worse than hiding them */
+        *(output_buffer + flatstringpos) = ',';
+        flatstringpos++;
+    }
+    list_iterator_destroy(attrvallist_itr);
+
+    /* kill that last comma or make an empty string*/
+    if (len)
+        *(output_buffer + flatstringpos - 1) = '\0';
+    else
+        *output_buffer = '\0';
+
+    return(0); 
+
+    cleanup:
+    if (avlist)
+        list_destroy(avlist);
+    if (attrvallist_itr)
+        list_iterator_destroy(attrvallist_itr);
+    if (avstring)
+        free(avstring);
+    return(1);
+}
+
+static void
+compress_hosts(genders_t gp)
+{
+    char **nodes;
+    char *attrsvals;
+    char **attrs, **vals;
+    int attrslen, valslen;
+    int attrscount;
+    int numattrs;  
+    int nodeslen;
+    int nodescount;
+    int numnodes; 
+    int maxattrsvalslen;
+    hash_t hattr = NULL; 
+    hash_t hrange = NULL; 
+    List hlist = NULL;
+    struct store_hostrange_data shd;
+    int i;
+
+    /* Similar to compress.  For each host, gather attrs & vals as a set,
+       compress hostlists of hosts with common sets.  The result is that each
+       node only appears only on one line.  */
+
+    /* need to treat values w/ raw inputs in order to compress */
+    if (genders_set_flags(gp, GENDERS_FLAG_RAW_VALUES) < 0)
+        _gend_error_exit(gp, "genders_set_flags");
+
+    if ((numnodes = genders_getnumnodes(gp)) < 0)
+        _gend_error_exit(gp, "genders_getnumnodes");
+
+    if ((numattrs = genders_getnumattrs(gp)) < 0)
+        _gend_error_exit(gp, "genders_getnumattrs");
+
+    if ((attrslen = genders_attrlist_create(gp, &attrs)) < 0)
+        _gend_error_exit(gp, "genders_attrlist_create");
+    
+    if ((valslen = genders_vallist_create(gp, &vals)) < 0)
+        _gend_error_exit(gp, "genders_vallist_create");
+
+    /* casses: I found that there are usually around numattrs/10 unique attrval
+     * sets in my examples, so numattrs should be plenty of bins.
+     * numattrs + 1, in case numattrs == 0
+     */
+    if (!(hattr = hash_create((numattrs + 1),
+                              (hash_key_f)hash_key_string,
+                              (hash_cmp_f)strcmp,
+                              _hosts_data_del))) {
+        fprintf(stderr, "hash_create: %s\n", strerror(errno));
+        exit(1);
+    }
+
+    if ((nodeslen = genders_nodelist_create(gp, &nodes)) < 0)
+        _gend_error_exit(gp, "genders_nodelist_create");
+
+    if ((nodescount = genders_getnodes(gp, nodes, nodeslen, NULL, NULL)) < 0)
+        _gend_error_exit(gp, "genders_getnodes");
+
+    /* Assume that each attr and val is max possible length, leave room for '='
+     * and (',' or terminating '\0').  */
+    maxattrsvalslen = (genders_getmaxattrlen(gp) + 2
+                       + genders_getmaxvallen(gp)) * numattrs;
+    if (!(attrsvals = (char *) malloc(maxattrsvalslen * sizeof(char)))) {

```

Comment:
If someone had a big genders database, this could integer overflow.  I think it's ok to be lazy here (I would be too), but perhaps add a comment to that affect. "Being lazy about integer overflow possibility".